### PR TITLE
copr: add vectorized compress and uncompress

### LINKED ISSUE #6775: UCP: Migrate scalar function `Compress` from TiDB
## Description

Port the scalar function `Compress` from TiDB to coprocessor.

## Score 

* 50

## Mentor(s)

* @breeswish 

## Recommended Skills

* Rust programming 

## Learning Materials

Already implemented expressions ported from TiDB

- https://github.com/tikv/tikv/tree/master/components/tidb_query/src/rpn_expr) 
- https://github.com/tikv/tikv/tree/master/components/tidb_query/src/expr)

### LINKED ISSUE #6778: UCP: Migrate scalar function `Uncompress` from TiDB
## Description

Port the scalar function `Uncompress` from TiDB to coprocessor.

## Score 

* 50

## Mentor(s)

* @sticnarf 

## Recommended Skills

* Rust programming 

## Learning Materials

Already implemented expressions ported from TiDB

- https://github.com/tikv/tikv/tree/master/components/tidb_query/src/rpn_expr) 
- https://github.com/tikv/tikv/tree/master/components/tidb_query/src/expr)

### PR DESCRIPTION
**Pull request by yeshengm**
### What problem does this PR solve?

Issue Number: close #6775 close #6778 

Problem Summary: add vectorized `compress` and `uncompress`.

### What is changed and how it works?

- implemented vectorized `compress` and `uncompress`
- factored out common functionality to the shared expression module

### Check List 

- Unit test

### CODE REVIEW

**Comment by yeshengm** [comment]
cc @breeswish

File: components/tidb_query_shared_expr/src/encryption.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
+use flate2::read::{ZlibDecoder, ZlibEncoder};
+use flate2::Compression;
+use std::io::Read;
+use tidb_query_datatype::codec::data_type::*;
+
+pub fn compress(input: &[u8]) -> Option<Bytes> {
```
Comment (sticnarf) [inline components/tidb_query_shared_expr/src/encryption.rs:6]: I think it better for the method and the `uncompress` method to return a `Result`.

**Comment by breeswish** [comment]
@iosmanthus @SunRunAway PTAL, thanks

File: components/tidb_query_shared_expr/src/encryption.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    }
+    let mut e = ZlibEncoder::new(input, Compression::default());
+    // preferred capacity is input length plus four bytes length header and one extra end "."
+    // max capacity is isize::max_value(), or will panic with "capacity overflow"
+    let mut vec = Vec::with_capacity((input.len() + 5).min(isize::max_value() as usize));
+    vec.resize(4, 0);
+    LittleEndian::write_u32(&mut vec, input.len() as u32);
+    match e.read_to_end(&mut vec) {
+        Ok(_) => {
+            // according to MySQL doc: append "." if ends with space
```
Comment (iosmanthus) [inline components/tidb_query_shared_expr/src/encryption.rs:20]: Could we add a doc link here?

**Comment by Xuanwo** [comment]
@yeshengm Hello, how's it going? Is there anything I can do to help?

**Comment by yeshengm** [comment]
@Xuanwo I'm not having the bandwidth to change this diff to ship-ready status. Feel free to commandeer this diff. Thanks!

**Comment by wangggong** [comment]
@Xuanwo @yeshengm This feature is added in #9141 , maybe this PR can be closed?

In the end, @yeshengm thanks a lot!

**Comment by skyzh** [comment]
Thanks for your contribution! However, https://github.com/tikv/tikv/pull/9141 has already implemented compress and uncompress. I'm afraid I should close this PR for now.

**Pull request closed without merging**